OCEANWATER-1073: PLEXIL interface to telemetry

kmdalal opened the pull request:
### Summary

This PR implements OW-1073 (Plexil interface to OWLAT telemetry).  It includes a fair amount of code refactoring, which was mainly a separation of OceanWATERS, OWLAT, and common code.  Many ROS topics and messages were renamed to follow the unified specification.  

Some misc unrelated changes are included, such as renaming identifiers to follow our coding guidelines, small improvements and refactoring of existing code, and updating of code comments.

### Reviewer Suggestions

@anthonykoutroulis : OceanWATERS build and test, review C++ code.
@AstroStucky : OWLAT build and test, review PLEXIL and C++ code

All reviewers, feel free to review and test anything!

### Notes on Code

- The target branch `releases/release-11-1` is an intermediate branch based on `master`.  The intention is to keep this work independent of any `ow_autonomy` changes that are for post-release 11.1, which would be on `noetic-devel`.
- The C++ code is still ripe for much refactoring, in particular templatization.  Leaving for future work, but feel free to make suggestions.
- The PLEXIL code could also be further refactored and better organized, also leaving for future work, but feel free to make suggestions.
- OWLAT's ROS action files are in `/opt/ros/noetic/share/owl_msgs`.  OceanWATERS' are in `ow_simulator/owl_msgs`.

### Build

Do a clean build of your workspace with `ow_autonomy` on this branch, and all other repositories on the `master` branch.  Why `master` and not `noetic-devel`?  Because this will eventually go into a "mini-release" 11.1 that updates only `ow_autonomy`.

Recall that ow_plexil is now built for _either_ OceanWATERS or OWLAT, not both.

OceanWATERS build: `catkin clean` followed by `catkin build`
OWLAT build: `catkin clean` followed by `catkin build -DOWLAT=ON`

### OWLAT test

You must have OWLAT Sim 5.3 installed.

1. Start OWLAT Sim
    See the first section in its _new_ (5.3) user guide.  Quick summary, each in a separate shell:
      a. `roscore`
      b. `roslaunch owl_msgs autonomy_interface.launch` (preceded by `source /opt/ros/noetic/setup.bash`)
      c. `launch_owlat Scoop CAM_FORCE_MOCK:true PT_USE_MOCK:true` (preceded by starting Python 3 virtual environment).

 
2. Start the OWLAT PLEXIL executive:

    `roslaunch ow_plexil owlat_exec.launch`

3. Run the plan `TestTelemetry`, which tests testbed-common telemetry and completes almost instantly.
    Use RQT's Plexil Plan Selection GUI, or enter `TestTelemetry.plx` at the prompt.

The output should look approximately like this attached log:
[owlat-telemetry.log](https://github.com/nasa/ow_autonomy/files/11763696/owlat-telemetry.log)

Note that the `nan` values for battery telemetry are expected, because the power model is not connected in this version of OWLAT.

4. Run the plan `TestOwlatTelemetry`, which tests all OWLAT-specific topics.  It completes in a few seconds.  Watch the lander animation window, but do not move or resize it during the test because this will botch the test.  You should see the end effectors change quickly.  The output should be as in this attached log:
[TestOwlatTelemetry.log](https://github.com/nasa/ow_autonomy/files/11710665/TestOwlatTelemetry.log)

### OceanWATERS test

1. Start the Atacama world.
    If you like you may increase the realtime factor in Gazebo to make tests go faster.

2. Start the PLEXIL executive:
    `roslaunch ow_plexil ow_exec.launch`

3. Run two plans back to back in the same simulator session.  You may select them in the RQT GUI or type in their filenames, adding the `.plx` extension.

  a) `TestTelemetry`.  This should complete almost instantly, and the output should be close to what's shown in this attached log:
[TestTelemetry.log](https://github.com/nasa/ow_autonomy/files/11710681/TestTelemetry.log)

  b) `ReferenceMission1`, for regression.  Success should be self-evident based on your familiarity with it.

AstroStucky commented:
OWLAT Test - Step 3 does not agree with the provided log. All the battery telemetry shows `nan`. Attached is the log in case you spot any other inconsistencies. I ran the `TestTelemetry` plan a second time and the battery values are still `nan` in the second run. 
[OwlatTestTelemetry-TS.log](https://github.com/nasa/ow_autonomy/files/11763646/OwlatTestTelemetry-TS.log)

Step 4 passed.

kmdalal commented:
> OWLAT Test - Step 3 does not agree with the provided log. All the battery telemetry shows `nan`.

Oh, crap.  The attached log file is messed up and contains _both_ the OWLAT and OW run.  I'm about to fix it.

The NAN values for battery are expected in OWLAT because they didn't connect the power model in this release.  I forgot to mention this in the PR, but will in a moment.

AstroStucky commented:
Ah, that clarifies things. Then the OWLAT test passes ✅

kmdalal commented:
> Ah, that clarifies things. Then the OWLAT test passes ✅

I have fixed the log, and noted the expected `nan` values.

AstroStucky commented on ow_plexil/src/plexil-adapter/joint_support.h:
```diff
@@ -15,6 +15,11 @@
 
 const size_t NumJoints = 9;
 
+// Supports both OceanWATERS lander (6 joints) and OWLAT (7 joints).
+// Many message files define a 7-element array for both testbeds; the
+// 7th position will be ignored for OceanWATERS.
```
This was intended from our unification document, but not the case in how it was implemented. See the wiki for details. 
https://github.com/nasa/ow_simulator/wiki/Lander-Simulation#arm-joint-telemetry

kmdalal replied:
I don't understand this comment.  Everything looks consistent to me...

AstroStucky replied:
Given that the 7th position is shoulder yaw, I don't think we should say it is ignored by OceanWATERS.

kmdalal replied:
I see my confusion now.  I have removed this comment, and updated the rest of the comments in this file.

AstroStucky reviewed (commented):
Code looks good as is, I just left a few code improvement suggestions, and commented on some ares that look like potential mistakes to me, but that may just be my lack of understanding of the codebase.

kmdalal commented on ow_plexil/src/plexil-adapter/OwInterface.cpp:
```diff
@@ -264,16 +175,6 @@ void OwInterface::jointStatesCallback
-      publish ("TiltDegrees", m_currentTiltRadians * R2D);
-    }
     JointTelemetries[i] = JointTelemetry {position, velocity, effort};
     publish ("JointPosition", position, i);
     publish ("JointVelocity", velocity, i);
```
This callback covers _all_ joints, including antenna.  It informs Lookups that are parameterized on the joint index, e.g. `Lookup(JointPosition(GRINDER))`.

kmdalal commented on ow_plexil/src/plans/owlat_plans/TestShearBevameter.plp:
```diff
@@ -3,21 +3,22 @@
+
+// Fails during bevameter action.  For some reason, the MaxTorque
+// threshold is never reached, and we've tried values as high as 1.0.
+// Reproducable at Ames, but not JPL.  We have given up as of 6/9/23.
```
That's right.  I've debugged this in detail with JPL and their best guess is that there's an issue with our DARTS installation.  It's not clear how to address this.  Given that no OceanWATERS users use this tool, we are giving up on this for now.

kmdalal commented on ow_plexil/src/plexil-adapter/LanderAdapter.cpp:
```diff
@@ -196,18 +196,106 @@ static void pan_tilt_move_joints (Command* cmd, AdapterExecInterface* intf)
+  v.resize(7);
+  vector<double> pose = LanderAdapter::s_interface->getArmPose();
+  for (size_t i = 0; i < 7; i++) {
+    v[i] = pose[i];  // implicit construction of Value from double
```
Good suggestion, done.

kmdalal commented on ow_plexil/src/plexil-adapter/LanderAdapter.cpp:
```diff
@@ -196,18 +196,106 @@ static void pan_tilt_move_joints (Command* cmd, AdapterExecInterface* intf)
+  v.resize(6);
+  vector<double> ft = LanderAdapter::s_interface->getArmEndEffectorFT();
+  for (size_t i = 0; i < 6; i++) {
+    v[i] = ft[i];  // implicit construction of Value from double
```
Done.

kmdalal commented on ow_plexil/src/plexil-adapter/LanderInterface.cpp:
```diff
@@ -50,56 +50,273 @@ static vector<string> LanderOpNames = {
+void LanderInterface::armJointAccelCb
+(const owl_msgs::ArmJointAccelerations::ConstPtr& msg)
+{
+  copy(msg->value.begin(), msg->value.end(), m_arm_joint_accelerations.begin());
```
Good catch.  I have added loops that publish the arm telemetry for each index, since the Lookups read values for just one joint.  I think the publish rate is slow enough for this not to pose a performance hit.

kmdalal commented on ow_plexil/src/plexil-adapter/OwlatInterface.cpp:
```diff
@@ -56,128 +56,82 @@ OwlatInterface* OwlatInterface::instance ()
+  m_subscribers.push_back
+    (make_unique<ros::Subscriber>
+     (m_genericNodeHandle ->
+      subscribe("/owlat_sim/ARM_TOOL", QueueSize,
```
This version of OWLAT sim supports only the old topic and message.  I have added comments here and in the header.

kmdalal commented:
@AstroStucky I have addressed all your comments so far.  I also removed some unused telemetry code, including the three 'publish' calls you commented on.   Please redo the OWLAT test, and let me know if there's anything else needing attention.

kmdalal commented:
@anthonykoutroulis I have made some significant code changes.  Could you please retry the OceanWATERS test?

AstroStucky reviewed (approved):
This all looks great! Still think that comment about the 7th value should be changed or removed. Happy to clarify further if needed. Approving for now so development can move along.

Great work!

anthonykoutroulis commented:
Re-tested and re-reviewed. Looks good!
